### PR TITLE
The inaugural commit for Leeloo's firmware.

### PR DESCRIPTION
**Pull request by ClicketySplit**
## Initial Configuration Files for Leeloo's Firmware

Modelled the configuration files and code for Leeloo's firmware based on the Lily58 and Corne split keyboards.  Slight variations, based on the inclusion of rotary encoders, and some wiring differences.

## Types of Changes
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [X] Documentation

## Issues Fixed or Closed by This PR

* This is a new PR.

## Checklist
- [ Y ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ Y ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ N ] My change requires a change to the documentation.
- [ N/A ] I have updated the documentation accordingly.
- [ Y ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ N/A ] I have added tests to cover my changes.
- [ Y ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

## Compile Results
qmk_build_results

## Fingers Crossed
This is my first contribution, so I am hoping that this is a good start.  I am open to feedback which may make this better.  Thank you for considering this new keyboard configuration.

### CODE REVIEW

File: keyboards/clickety_split/leeloo/keymaps/default/config.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+// Select hand configuration
+#define MASTER_LEFT
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/keymaps/default/config.h:21]: This is the default behavior, and is not needed. 
```suggestion
// #define MASTER_LEFT
```

File: keyboards/clickety_split/leeloo/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+#include "leeloo.h"
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/keymaps/default/keymap.c:19]: Not needed, as that is what the QMK_KEYBOARD_H is defined as, actually. 
```suggestion
```

File: keyboards/clickety_split/leeloo/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    } else {
+        // Version Information
+        render_secondary_oled();
+    }
+
+    return false;
+
+}
+
+#endif
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/keymaps/default/keymap.c:201]:
```suggestion
```

File: keyboards/clickety_split/leeloo/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+            tap_code(KC_PGDN);
+        } else {
+            tap_code(KC_PGUP);
+        }
+    }
+
+    return false;
+
+}
+#endif
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/keymaps/default/keymap.c:225]:
```suggestion
```

File: keyboards/clickety_split/leeloo/leeloo.c [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "leeloo.h"
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/leeloo.c:18]: This allows for a default behavior to be defined, without the need for the user to specify the behavior.  This is very useful for the configurator which doesn't have a way to configure this behavior, currently.


```suggestion
#include "leeloo.h"

#ifdef OLED_ENABLE
oled_rotation_t oled_init_kb(oled_rotation_t rotation) {
    if (!is_keyboard_master()) {
        return OLED_ROTATION_180;
    } else {
        return rotation;
    }
}

static void render_keylock_status(uint8_t led_usb_state) {
    oled_write_P(PSTR("Lock: "), false);
    oled_write_P(PSTR("CAPS"), led_usb_state & (1 << USB_LED_CAPS_LOCK));
    oled_write_P(PSTR(" "), false);
    oled_write_P(PSTR("NUML"), led_usb_state & (1 << USB_LED_NUM_LOCK));
    oled_write_P(PSTR(" "), false);
    oled_write_ln_P(PSTR("SCLK"), led_usb_state & (1 << USB_LED_SCROLL_LOCK));
}

static void render_mod_status(uint8_t modifiers) {
    oled_write_P(PSTR("Mods: "), false);
    oled_write_P(PSTR("Sft"), (modifiers & MOD_MASK_SHIFT));
    oled_write_P(PSTR(" "), false);
    oled_write_P(PSTR("Ctl"), (modifiers & MOD_MASK_CTRL));
    oled_write_P(PSTR(" "), false);
    oled_write_P(PSTR("Alt"), (modifiers & MOD_MASK_ALT));
    oled_write_P(PSTR(" "), false);
    oled_write_P(PSTR("GUI"), (modifiers & MOD_MASK_GUI));
}

static void render_secondary_oled(void) {
    // Version Information
    oled_write_P(PSTR("Leeloo\n\n"), false);
    oled_write_P(PSTR("Firmware: "), false);
    oled_write_P(PSTR(LEELOO_VERSION), false);
    oled_write_P(PSTR("\n"), false);
    oled_write_P(PSTR("Clickety Split Ltd."), false);
}

static static void render_status(void) {
    // Host Keyboard Layer Status
    switch (get_highest_layer(default_layer_state)) {
        case _BASE:
            oled_write_P(PSTR("QWERTY | "), false);
            break;
    }

    // Host Keyboard Layer Status
    switch (get_highest_layer(layer_state)) {
        case 0:
            oled_write_P(PSTR("Base   \n"), false);
            break;

        case _LOWER:
            oled_write_P(PSTR("Lower   \n"), false);
            break;

        case _RAISE:
            oled_write_P(PSTR("Raise   \n"), false);
            break;

        case _ADJUST:
            oled_write_P(PSTR("Adjust  \n"), false);
            break;

        default:
            oled_write_P(PSTR("Unknown \n"), false);
    }

    oled_write_P(PSTR("\n"), false);
    render_keylock_status(host_keyboard_leds());
    render_mod_status(get_mods() | get_oneshot_mods());
}

bool oled_task_kb(void) {
    if (!oled_task_user()) {
        return false; 
    }

    if (is_keyboard_master()) {
        // Renders the current keyboard state (layer, lock, caps, scroll, etc)
        render_status();
    } else {
        // Version Information
        render_secondary_oled();
    }

    return false;

}
#endif // OLED_ENABLE

#ifdef ENCODER_ENABLE
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) { return false; }
    if (index == 0) {
        // Volume control
        if (clockwise) {
            tap_code_delay(KC_VOLD, 10);
        } else {
            tap_code_delay(KC_VOLU, 10);
        }
    } else if (index == 1) {
        // Page up/Page down
        if (clockwise) {
            tap_code(KC_PGDN);
        } else {
            tap_code(KC_PGUP);
        }
    }
    return true;
}
#endif
```

File: keyboards/clickety_split/leeloo/leeloo.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "quantum.h"
+
+#if defined(KEYBOARD_clickety_split_leeloo_rev1)
+    #include "rev1.h"
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/leeloo.h:23]:
```suggestion
#    include "rev1.h"
```

File: keyboards/clickety_split/leeloo/rev1/config.h [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x1004
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/rev1/config.h:23]: This vendor ID is already in use, and needs to be changed. 

Anything about 0x4000 should be okay, and the higher, the better.

File: keyboards/clickety_split/leeloo/rev1/config.h [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+#define PRODUCT_ID      0x2022
+#define DEVICE_VER      0x0013
+#define MANUFACTURER    Clickety Split Ltd
+#define PRODUCT         Leeloo
+#define LEELOO_VERSION  "v1.0"
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/rev1/config.h:28]:
```suggestion
```

File: keyboards/clickety_split/leeloo/rev1/config.h [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+#define MATRIX_COLS 6
+
+// wiring of each half
+#define MATRIX_ROW_PINS { C6, D7, E6, B4, B5 }
+#define MATRIX_COL_PINS { F6, F7, B1, B3, B2, B6 }
+#define DIODE_DIRECTION COL2ROW
+
+/* communication between sides */
+#define USE_SERIAL
+#define SERIAL_USE_MULTI_TRANSACTION
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/rev1/config.h:42]: default behavior.
```suggestion
```

File: keyboards/clickety_split/leeloo/rev1/config.h [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCE 5
+
+/* encoder support */
+#define ENCODERS_PAD_A { F5 }
+#define ENCODERS_PAD_B { F4 }
+#define ENCODERS_PAD_A_RIGHT { F4 }
+#define ENCODERS_PAD_B_RIGHT { F5 }
+#define ENCODER_RESOLUTION 2
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/rev1/config.h:56]: Shold be set by user. 
```suggestion
```

File: keyboards/clickety_split/leeloo/rev1/config.h [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+/* disable action features */
+//#define NO_ACTION_LAYER
+//#define NO_ACTION_TAPPING
+//#define NO_ACTION_ONESHOT
+#define NO_ACTION_MACRO
+#define NO_ACTION_FUNCTION
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/rev1/config.h:63]: Not needed. 
```suggestion
```

File: keyboards/clickety_split/leeloo/rules.mk [hunk 12]
```diff
@@ -0,0 +1,36 @@
+# MCU name
+MCU = atmega32u4
+
+# Bootloader selection
+BOOTLOADER = caterina       # Pro Micro
+#BOOTLOADER = atmel-dfu		# Elite-C
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/rules.mk:6]:
```suggestion
```

File: keyboards/clickety_split/leeloo/rules.mk [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no      # Breathing sleep LED during USB suspend
+
+# To enable Rotary Encoders
+ENCODER_ENABLE = no
+
+# To enable OLED Displays
+OLED_ENABLE = no
+OLED_DRIVER = SSD1306      # OLED display
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/rules.mk:33]:
```suggestion
NKRO_ENABLE = no           # Enable N-Key rollover
BACKLIGHT_ENABLE = no      # Enable keyboard backlight functionality
AUDIO_ENABLE = no          # Audio output
RGBLIGHT_ENABLE = no       # Enable WS2812 RGB underlight.
```

File: keyboards/clickety_split/leeloo/rules.mk [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+MCU = atmega32u4
+
+# Bootloader selection
+BOOTLOADER = caterina       # Pro Micro
+#BOOTLOADER = atmel-dfu		# Elite-C
+
+# Build Options
+#   change to "no" to disable the options, or define them in the Makefile in
+#   the appropriate keymap folder that will get included automatically
+#
```
Comment (drashna) [inline keyboards/clickety_split/leeloo/rules.mk:11]:
```suggestion
#   change to "no" to disable the options
#
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/clickety_split/leeloo/rev1/config.h (hunk 7 above)
Comment (ClicketySplit) [reply keyboards/clickety_split/leeloo/rev1/config.h]: Hi @drashna!

I was able to update the VENDOR_ID to 0x4013.  I hope this ID will be okay. :)

**Comment by ClicketySplit** [comment]
Hi @drashna,

Thank you very much for taking the time to review the PR, and guiding me through the process.  I really appreciate it!

:)

**Review by drashna — APPROVED** [review]

File: keyboards/clickety_split/leeloo/rules.mk [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+EXTRAKEY_ENABLE = no       # Audio control and System control
+CONSOLE_ENABLE = no        # Console for debug
+COMMAND_ENABLE = no        # Commands for debug and configuration
+NKRO_ENABLE = no           # Enable N-Key rollover
+BACKLIGHT_ENABLE = no      # Enable keyboard backlight functionality
+AUDIO_ENABLE = no          # Audio output
+RGBLIGHT_ENABLE = no       # Enable WS2812 RGB underlight.
+
+SPLIT_KEYBOARD = yes
+DEFAULT_FOLDER = clickety_split/leeloo/rev1
```
Comment (fauxpark) [inline keyboards/clickety_split/leeloo/rules.mk:21]:
```suggestion
BOOTLOADER = caterina

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = no       # Enable Bootmagic Lite
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = no        # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
NKRO_ENABLE = no            # Enable N-Key rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
SPLIT_KEYBOARD = yes

DEFAULT_FOLDER = clickety_split/leeloo/rev1
```

File: keyboards/clickety_split/leeloo/config.h [hunk 16]
```diff
@@ -0,0 +1,20 @@
+/* Copyright 2022 Clickety Split Ltd.
```
Comment (fauxpark) [inline keyboards/clickety_split/leeloo/config.h:1]: This file can be deleted.

File: keyboards/clickety_split/leeloo/keymaps/default/readme.md [hunk 17]
```diff
@@ -0,0 +1,49 @@
+# Clickety Split | Leeloo
```
Comment (fauxpark) [inline keyboards/clickety_split/leeloo/keymaps/default/readme.md:1]: This file can be deleted.

File: keyboards/clickety_split/leeloo/leeloo.h [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "quantum.h"
+
+enum layer_names {
```
Comment (fauxpark) [inline keyboards/clickety_split/leeloo/leeloo.h:22]: This should go in the keymap.c.

File: keyboards/clickety_split/leeloo/readme.md [hunk 19]
```diff
@@ -0,0 +1,49 @@
+# Clickety Split | Leeloo
```
Comment (fauxpark) [inline keyboards/clickety_split/leeloo/readme.md:1]: Please stick to the readme template.

File: keyboards/clickety_split/leeloo/keymaps/default/rules.mk [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+# To enable OLED Displays
+OLED_ENABLE = no
+OLED_DRIVER = SSD1306       # OLED display
+
+# Bootloader selection
+BOOTLOADER = caterina       # Pro Micro
+#BOOTLOADER = atmel-dfu		# Elite-C
+
+# Enable Link Time Optimization
+LTO_ENABLE = yes			# Enable help with smaller firmware size
```
Comment (fauxpark) [inline keyboards/clickety_split/leeloo/keymaps/default/rules.mk:19]:
```suggestion
LTO_ENABLE = yes
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by ClicketySplit** [comment]
Hi @fauxpark!

Thank you very much for taking the time to review this PR and to provide feedback and recommendations.  Hopefully the changes I have made are appropriate for your review.

:)

File: keyboards/clickety_split/leeloo/leeloo.c [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+        case _ADJUST:
+            oled_write_P(PSTR("Adjust  \n"), false);
+            break;
+
+        default:
+            oled_write_P(PSTR("Unknown \n"), false);
+    }
+
+    oled_write_P(PSTR("\n"), false);
+    render_keylock_status(host_keyboard_leds());
```
Comment (fauxpark) [inline keyboards/clickety_split/leeloo/leeloo.c:89]:
```suggestion
    render_keylock_status(host_keyboard_led_state());
```

File: keyboards/clickety_split/leeloo/leeloo.c [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+    }
+}
+
+static void render_keylock_status(uint8_t led_usb_state) {
+    oled_write_P(PSTR("Lock: "), false);
+    oled_write_P(PSTR("CAPS"), led_usb_state & (1 << USB_LED_CAPS_LOCK));
+    oled_write_P(PSTR(" "), false);
+    oled_write_P(PSTR("NUML"), led_usb_state & (1 << USB_LED_NUM_LOCK));
+    oled_write_P(PSTR(" "), false);
+    oled_write_ln_P(PSTR("SCLK"), led_usb_state & (1 << USB_LED_SCROLL_LOCK));
```
Comment (fauxpark) [inline keyboards/clickety_split/leeloo/leeloo.c:35]:
```suggestion
static void render_keylock_status(led_t led_state) {
    oled_write_P(PSTR("Lock: "), false);
    oled_write_P(PSTR("CAPS"), led_state.caps_lock);
    oled_write_P(PSTR(" "), false);
    oled_write_P(PSTR("NUML"), led_state.num_lock);
    oled_write_P(PSTR(" "), false);
    oled_write_ln_P(PSTR("SCLK"), led_state.scroll_lock);
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by ClicketySplit** [comment]
Hi @fauxpark!

Thank you for recommending the two changes; I was able to test the requests and they work great!  I learned something new!

:)

**Review by fauxpark — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!

**Comment by ClicketySplit** [comment]
Thank you both for all your help! I learned a lot from your feedback!

:)

------- Original Message -------

**Pull request merged**